pac4j-cas/src/main/java/org/pac4j/cas/client/rest/CasRestAuthenticator.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2012 - 2015 pac4j organization
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.pac4j.cas.client.rest;
+
+import org.apache.commons.httpclient.HttpStatus;
+import org.jasig.cas.client.validation.TicketValidator;
+import org.pac4j.http.credentials.UsernamePasswordCredentials;
+import org.pac4j.http.credentials.authenticator.Authenticator;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * This is {@link CasRestAuthenticator}.
+ *
+ * @author Misagh Moayyed
+ * @since 1.8.0
+ */
+public class CasRestAuthenticator implements Authenticator<UsernamePasswordCredentials> {
+    private final URL endpointURL;
+    private final TicketValidator ticketValidator;
+
+    public CasRestAuthenticator(final URL endpointURL, final TicketValidator ticketValidator) {
```

Comment:
I understand your opinion and value it. It would make sense for the LDAP implementation if I rely on ldaptive and an `Authenticator`. Because it's a very complex object I don't really want to replace it (so much work).

The philosophy of `pac4j` is to hide and abstract as much as possible from internal libraries. You should never know that it uses the `cas-client-core` library for Java or `Scribe` for OAuth support.
In the future, if `pac4j` becomes internally reactive (I had one more discussion with Tim Fox, the creator of Vertx about that), we may change the underlying library.

Your argument (be able to switch from Cas30 to Cas40) is relevant but in practice, it doesn't work: if ever we upgrade to Cas40, I'm pretty sure we'll need to change the internal logic as well. So it doesn't really make sense to offer this ability.
The last update to version 3.3.3 of the CAS client generates breaking changes: upgrading is not that seamless as well.

So I'd really prefer to set the CAS server prefix url and the CAS REST url and hide the ticket alidator.